Add CLI config in Cargo.toml to set feature flags

TimJentzsch opened the pull request:
# Objective

Closes #250.

Bevy apps often want to configure different features in dev vs. release mode or when targeting native vs. web.
With native `cargo`, that's currently a pain to do.

With this PR, the user can configure `features` and `default_features` in the CLI in general, but also specific for the different modes as mentioned above.

# Solution

The `cargo metadata` output also includes the contents of the `metadata` table of each package.
We retrieve the value for the package that we build for and then extract the config from that metadata.

For example, a user can configure `package.metadata.bevy_cli.web.release` to set up features when compiling for the web in release mode.
The `bevy_cli.release` and `bevy_cli.web` configs are hereby merged into the more specific configs to make configuration a bit less boilerplaty.

Note that the workspace config is currently _not_ considered, because configuring features only makes sense for packages.

# TODO

- Determine package for the binary that needs to be compiled
- Apply config to build/run arguments

# Feedback wanted

- Is the config merge order base -> profile -> platform -> platform-profile the one you'd expect?

BD103 commented:
> The `cargo metadata` output also includes the contents of the `metadata` table of each package.

Wow, I didn't know that! The linter could probably use this to simplify our config loading!

TimJentzsch commented:
It now works!

Well, at least for `web` builds. For `native` builds we currently ignore the config, because to read the config, we need to know the package to read the metadata from.
And to know the package, we need to run `cargo build`/`cargo run`. But that's where we want to pass the feature flags from the config, a circular dependency!

With the web commands we get around this problem, because we need to manually determine the targets we want to build anyway. However, this doesn't have exactly the same behavior as `cargo`. For example, `cargo build` can build multiple targets, but we currently only support one target at a time with `bevy build web`.
For native I think it's best if we keep full compatibility with `cargo` to avoid weird bugs or unexpected behavior...

TimJentzsch commented:
For native, I now also use our logic to determine the binary we want to run.
But the binary is not applied to the `cargo` command itself (in order to preserve behavior), it's only used to determine the package where we get the CLI config from.

I think that's the best of both worlds for now: We still get the config for native, but we also maintain compatibility with `cargo` for the native commands.

DaAlbrecht reviewed (changes requested):
I like the solution a lot, feels very intuitive and nice!! Would also be nice to support `rustflags` in a future PR.

I think this needs a section in the `README.md`

TimJentzsch commented on src/bin_target.rs:
```diff
@@ -1,12 +1,18 @@
 #[derive(Debug, Clone)]
 pub struct BinTarget {
+    /// The package containing the binary.
+    pub package: Package,
```
I thought this would introduce complications when `BinTarget` needs a lifetime, but it actually played out very nicely.
Thanks!

TimJentzsch commented:
I added a section to the README (nice catch!) and implemented the package as reference change (also worked out nicely!).
Please let me know if I should change anything else :)

DaAlbrecht reviewed (approved):
top! Thanks so much 😊
